### PR TITLE
Track step "editing" state with a separate `WizardStep.editing` flag

### PR DESCRIPTION
**Pull request by earshinov**
Solves #213 by following the suggestion from https://github.com/madoar/angular-archwizard/issues/213#issuecomment-501008323

<a href="https://gitpod.io/#https://github.com/madoar/angular-archwizard/pull/221"></a>

### CODE REVIEW

**Comment by earshinov** [comment]
Note: Completion steps never receive "editing" state, because their (and wizard's) "completed" state is reset each time a completion step is left. So when a completion step is entered next time, it is impossible for it to be already completed.

File: src/lib/util/test-utils.ts [hunk 1]
```diff
@@ -20,9 +22,13 @@ export function checkWizardState(
...
     expect(step.selected).toBe(index === selectedStepIndex, `expected only step ${index} to be selected`);
 
-    // All steps before the selected step need to be completed
+    // Check completed step indexes
     expect(step.completed).toBe(completedStepIndexes.includes(index),
       `expected step ${index} ${completedStepIndexes.includes(index) ? 'to be completed' : 'not to be completed'}`);
+
+    // Check step "editing" state
+    const expectation = index === selectedStepIndex && selectedStepEditing ? 'to be in editing state' : 'not to be in editing state';
+    expect(step.editing).toBe(index === selectedStepIndex && selectedStepEditing, `expected step ${index} ${expectation}`);
```
Comment (madoar) [inline src/lib/util/test-utils.ts]: Shouldn't this also check the constraint `if selectedStepEditing then completedStepIndexes.contains(selectedStepIndex)`?

Comment (earshinov) [reply src/lib/util/test-utils.ts]: Good idea. Implemented in 2f354a5

**Comment by earshinov** [comment]
> Are these changes intended?
No, definitely not, although in the first case the new behavior seems expected to me (at first glance at least). I will try to sort it out.

**Comment by earshinov** [comment]
> Previously only the step to which the user went back to was marked as "editing". After the user left this step in forward direction the next step was marked as a "normal" (not "editing") step again. Now the wizard seems to remembers all steps the user entered at least once as completed and therefore marks them as "editing" upon reentering.

From what I see, the behavior is the opposite. The current 4.0.0 version remembers completed steps and marks the step as "editing". The `step-editing-state` version marks the step as normal. Repro steps

1. Open #/navigation-mode/semi-strict page for the 4.0.0 version, #/navigation-mode/configurable for the `step-editing-state` branch version
2. Press Continue to go to step 2
3. Press Continue to go to step 3
4. Navigate back to step 1 using the navigation bar
5. Both versions mark step 1 as "editing" - OK
6. Press Continue to go forward to step 2 again

4.0.0 version:

4 0 0 version

Branch version:

branch version

The branch version clears "completed" state (and `done` CSS class) for steps.

**Comment by earshinov** [comment]
Ah, sorry, in the branch version in order to properly emulate the "semi-strict" mode one first has to select "Navigate forward: Allow". In this case, the behavior of the wizard matches the behavior or v 4.0.0. @madoar , can you provide a detailed reproduction scenario?

**Comment by earshinov** [comment]
> (b.) previously in "strict" mode the user was only able to navigate to the next step in forward direction independent of whether he went back to a previous step or not. Now the user is able to go back to an later previously completed step, because the "chain" of completed steps is never broken.

Can you illustrate this case as well? Do you mean navigation using the navigation bar? From what I see, the chain is broken right after the user navigates back. Scenario:
1. Open http://localhost:4249/#/navigation-mode/default (the default navigation mode is equivalent to the old "strict" one)
2. Press Continue to go to step 2
3. Press Continue to go to step 3
4. Go back to step 1 using the navigatiom bar

navigation back in strict mode

The user is only allowed to navigate forward to the step that immediately follows the current one, and only with the Continue button. Not with the navigation bar. This behavior seems to match the behavior of v 4.0.0

**Comment by earshinov** [comment]
Here is the relevant code taken from `ConfigurableNavigationMode.transition`:

```typescript
  protected transition(wizard: WizardComponent, destinationIndex: number): void {
    if (this.navigateForward === 'deny') {
      // set all steps after the destination step to incomplete
      wizard.wizardSteps
        .filter((step, index) => wizard.currentStepIndex > destinationIndex && index > destinationIndex)
        .forEach(step => step.completed = false);
    }

    super.transition(wizard, destinationIndex);
  }
```

**Comment by madoar** [comment]
Sorry my bad I only looked at the code and didn't find the section where `step.completed = false` is executed.

The reason for my first issue was the modified test case:

https://github.com/madoar/angular-archwizard/blob/2f354a584d4c130eb29928dcbc65ab48e51b01f7/src/lib/navigation/wizard-navigation-allow-forward.spec.ts#L66-L90

When looking at the test it seems to be the case that the editing state is carried over to the next step.

**Comment by earshinov** [comment]
Actually, this is very good that you carefully looked through the tests, because in most cases I updated them almost mechanically.

**Comment by madoar** [comment]
Currently blocked by https://github.com/earshinov/extract-scss-variables/issues/1

**Comment by madoar** [comment]
@earshinov this seems to only partially fix the issue. Now when entering a completion step the step is not shown as completed:

grafik

**Comment by earshinov** [comment]
@madoar , It seems to be another issue. The question is which state should take greater priority: "done" or "completed". In the current version, the completion step receives "done", "completed" and "current" CSS classes, of which "current" takes priority.

The 4.0.0 version indeed renders the completion step indicator with green color, but are you sure it is the right behavior? In my opinion, the new behavior looks more correct, because (arguably) it is more important to indicate which step is current than whether a step is completed or not.

**Comment by madoar** [comment]
I agree that this is the correct behavior for a normal step. A completion step differs from this behavior because it is used to signify that the wizard has been completed. Therefore I think it is more correct to set the step to completed in this case.

**Comment by earshinov** [comment]
@madoar , I think I will create a separate pull request for this issue. It significantly complicates CSS styles, which also means that it will be more difficult for library users to write custom styles (should they need to do it). I guess you would like to contemplate a little bit on whether such complications are justified.

My suggestion is to leave the current `develop`-branch behavior as default. I don't think that many users will pay much attention to differences, and even less will strongly require one or the other version. And if they do require a version other than the default, they have all means to customize wizard's CSS.

**Comment by earshinov** [comment]
Here we go: https://github.com/madoar/angular-archwizard/pull/223

**Pull request merged**